dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/GitHubFileTree.java

```diff
@@ -0,0 +1,115 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package io.dockstore.webservice.helpers;
+
+import io.dockstore.common.Utilities;
+import io.dockstore.webservice.CustomWebApplicationException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.channels.SeekableByteChannel;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import org.apache.commons.compress.archivers.zip.ZipArchiveEntry;
+import org.apache.commons.compress.archivers.zip.ZipFile;
+import org.apache.commons.compress.utils.SeekableInMemoryByteChannel;
+import org.apache.commons.io.IOUtils;
+import org.apache.http.HttpStatus;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class GitHubFileTree implements FileTree {
+
+    private static final Logger LOG = LoggerFactory.getLogger(GitHubFileTree.class);
+
+    private final GitHubSourceCodeRepo gitHubSourceCodeRepo;
+    private final String repository;
+    private final String ref;
+    private final ZipFile zipFile;
+    private final Map<String, ZipArchiveEntry> pathToEntry;
+
+    public GitHubFileTree(GitHubSourceCodeRepo gitHubSourceCodeRepo, String repository, String ref) {
+        this.gitHubSourceCodeRepo = gitHubSourceCodeRepo;
+        this.repository = repository;
+        this.ref = ref;
+        // Read the Zip contents and create an in-memory ZipFile.
+        byte[] zipBytes = gitHubSourceCodeRepo.readZip(repository, ref);
+        LOG.info("downloaded Zip of GitHub repository: %n bytes".formatted(zipBytes.length));
+        SeekableByteChannel zipChannel = new SeekableInMemoryByteChannel(zipBytes);
+        try {
+            zipFile = new ZipFile(zipChannel);
+        } catch (IOException e) {
+            LOG.error("could not read zip archive of GitHub repository", e);
+            throw new CustomWebApplicationException("could not read GitHub repository", HttpStatus.SC_BAD_REQUEST);
+        }
+        // Create a Map of absolute paths to Zip file entries.
+        pathToEntry = Collections.list(zipFile.getEntries()).stream()
+            .filter(entry -> !entry.isDirectory() && !entry.isUnixSymlink())
+            .collect(Collectors.toMap(this::pathFromEntry, entry -> entry, (valueA, valueB) -> valueA));
+    }
+
+    @Override
+    public String readFile(String path) {
+        // If specified file is included in the ZipFile, uncompress its content, convert it to a string, and return it.
+        ZipArchiveEntry entry = pathToEntry.get(path);
+        if (entry != null) {
+            try (InputStream in = zipFile.getInputStream(entry)) {
+                return IOUtils.toString(in, StandardCharsets.UTF_8);
```

Comment:
What would happen if the file were binary? This would throw an IOException, in which case, all good?

Reply:
> What would happen if the file were binary? This would throw an IOException, in which case, all good?

This code should produce the same result as `GitHubSourceCodeRepo.readFile` would've - that is, it reads the bytes of the binary file, converts them to a UTF-8 string, and returns the result.  No exception thrown.